src/librustdoc/test.rs

```diff
@@ -308,49 +228,62 @@ fn run_test(
     };
     let output_file = outdir.path().join("rust_out");
 
-    let config = interface::Config {
-        opts: sessopts,
-        crate_cfg: config::parse_cfgspecs(cfgs),
-        input,
-        input_path: None,
-        output_file: Some(output_file.clone()),
-        output_dir: None,
-        file_loader: None,
-        diagnostic_output: DiagnosticOutput::Raw(box Sink(data.clone())),
-        stderr: Some(data.clone()),
-        crate_name: None,
-        lint_caps: Default::default(),
-    };
+    let current_dir = env::current_dir().expect("rustdoc needs to create a tempfile");
```

Comment:
needed to create the rustdoc input file in the dir where the file with the test is due to https://github.com/rust-lang/rust/issues/43153
maybe is not the good is it not common that when ci builds and test that the repo is readonly?

Reply:
Hm, I would expect us to be able to get away with the same behavior as before, where I believe we wrote everything to temporary directories... why can't we do the same here?

Reply:
the printout from the test is 
`---- [rustdoc] rustdoc/issue-43153.rs stdout ----

error: rustdoc failed!
status: exit code: 101
command: "/home/andjo403/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/home/andjo403/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/andjo403/rust/build/x86_64-unknown-linux-gnu/test/rustdoc/issue-43153/auxiliary" "-o" "/home/andjo403/rust/build/x86_64-unknown-linux-gnu/test/rustdoc/issue-43153" "/home/andjo403/rust/src/test/rustdoc/issue-43153.rs" "--test"
stdout:
------------------------------------------

running 1 test
test src/test/rustdoc/issue-43153.rs - Foo (line 6) ... FAILED

failures:

---- src/test/rustdoc/issue-43153.rs - Foo (line 6) stdout ----
"/home/andjo403/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--cfg" "rustdoc" "--cfg" "doctest" "--edition" "2015" "/tmp/rustdoctestlnftdr/input.rs" "-o" "/tmp/rustdoctestlnftdr/rust_out" "-L" "/home/andjo403/rust/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/andjo403/rust/build/x86_64-unknown-linux-gnu/test/rustdoc/issue-43153/auxiliary"
Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "error: couldn\'t read /tmp/rustdoctestlnftdr/auxiliary/empty.rs: No such file or directory (os error 2)\n --> /tmp/rustdoctestlnftdr/input.rs:2:1\n  |\n2 | include!(\"auxiliary/empty.rs\");\n  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n\nerror: aborting due to previous error\n\n" }
Couldn't compile the test.

failures:
    src/test/rustdoc/issue-43153.rs - Foo (line 6)

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out`

Reply:
Would it fix it if instead of making the tempdir be relative to the current_dir we set the current directory for the rustc command to be the same as rustdoc?

Reply:
when I looked in to it more it seems like the compile from stdin functionality in rustc was used, but I do not know how it works still needs to get the same path as to the file with the test but then at least no new files is created

Reply:
If we can use stdin successfully, then that works too. The important bit is to be 1-1 with what happened previously as much as possible, the specific mechanism doesn't matter too much.

Reply:
did not work with using the stdin as it is now got an ICE.
is it possible to add som hidden property to rustc to say that it is a doctest and the path to the file is something or maybe an env variable due to as it is now there is som special handling of e.g. line numbers in rustc when there is a FileName::DocTest that is missing with this solution

`echo 'include!("auxiliary/empty.rs"); pub fn main(){}' | rustc  -
thread 'rustc' panicked at 'cannot resolve relative path in non-file source `<anon>`', src/libsyntax/ext/source_util.rs:189:22
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.36.0 (a53f9df32 2019-07-03) running on x86_64-unknown-linux-gnu`

Reply:
Can you please create an issue for that ICE? Every ICE is a bug.

Reply:
ICE issue created https://github.com/rust-lang/rust/issues/63900

Reply:
I'll work on fixing said ICE and also providing a path forward here -- it's a larger patch and will probably need lang or compiler team sign off as an expansion of stable surface area. I'll post a PR soon.

Reply:
One problem that I've noted is that since we no longer have a way to thread the line offset through that'll lead to worse error messages, so we might need to add some surface area for that use case. I'm not sure what the best way to do it is (this PR will probably be stalled for some time until that's figured out, so I've labeled as S-blocked).

Reply:
yes have the same problem the only thing that I came up with was to add some env var like 
`env::var_os("RUSTDOC_TEST_FILE_AND_LINE")` in to rustc and if that is set crate a Input::Str(FileName::DocTest) from that like was done before instead of an Input::Str(FileName::Anon)

Reply:
Let's get tests here passing with an env variable like that (probably read [here](https://github.com/rust-lang/rust/blob/master/src/librustc_driver/lib.rs#L442)) though we'll want to name it something like `UNSTABLE_RUSTDOC_TEST_FILE_AND_LINE`.

Reply:
that is the location that I was thinking of also to add that.
will work on it tomorrow as it is late here now

Reply:
Yeah, the patch is looking like it'd basically be down to just threading the values through the environment so I won't actually post a separate PR for it -- it'd be good to have those changes in a separate commit for easier readability though.